roles/openshift_health_checker/library/aos_version.py

```diff
@@ -1,91 +1,189 @@
-available, and if the version requested is available down to the given
-precision.
-
-Multiple versions available suggest that multiple repos are enabled for the
-different versions, which may cause installation problems.
+Ansible module for yum-based systems determining if multiple releases
+of an OpenShift package are available, and if the release requested
+(if any) is available down to the given precision.
+
+For Enterprise, multiple releases available suggest that multiple repos
+are enabled for the different releases, which may cause installation
+problems. With Origin, however, this is a normal state of affairs as
+all the releases are provided in a single repo with the expectation that
+only the latest can be installed.
+
+Code in the openshift_version role contains a lot of logic to pin down
+the exact package and image version to use and so does some validation
+of release availability already. Without duplicating all that, we would
+like the user to have a helpful error message early on if things aren't
+going to work out right.
+
+TODO: fail gracefully on non-yum systems (dnf in Fedora)
 '''
 
 import yum  # pylint: disable=import-error
 
 from ansible.module_utils.basic import AnsibleModule
 
 
-def main():  # pylint: disable=missing-docstring,too-many-branches
+class AosVersionException(Exception):
+    '''Base exception class for package version problems'''
+    def __init__(self, message, problem_pkgs=None):
+        Exception.__init__(self, message)
+        self.problem_pkgs = problem_pkgs
+
+
+def main():
+    '''Entrypoint for this Ansible module'''
     module = AnsibleModule(
         argument_spec=dict(
-            prefix=dict(required=True),  # atomic-openshift, origin, ...
-            version=dict(required=True),
+            requested_openshift_release=dict(type="str", default=''),
+            openshift_deployment_type=dict(required=True),
+            rpm_prefix=dict(required=True),  # atomic-openshift, origin, ...?
         ),
         supports_check_mode=True
     )
 
-    def bail(error):  # pylint: disable=missing-docstring
-        module.fail_json(msg=error)
+    # determine the packages we will look for
+    rpm_prefix = module.params['rpm_prefix']
+    if not rpm_prefix:
+        module.fail_json(msg="rpm_prefix must not be empty")
+    expected_pkgs = set([
+        rpm_prefix,
+        rpm_prefix + '-master',
+        rpm_prefix + '-node',
+    ])
 
-    rpm_prefix = module.params['prefix']
+    # determine what level of precision the user specified for the openshift version.
+    # should look like a version string with possibly many segments e.g. "3.4.1":
+    requested_openshift_release = str(module.params.get('requested_openshift_release', ''))
+
+    # get the list of packages available and complain if anything is wrong
+    try:
+        pkgs = _retrieve_available_packages(expected_pkgs)
+        if requested_openshift_release:
```

Comment:
Aha, so the expected behavior is "if no version is specified, this check succeeds", right? I'm good with that. Since we're already improving the wording in the module docstring, maybe we can specify this behavior there also?